### PR TITLE
nd follow-up

### PR DESCRIPTION
**Pull request by soupault**
- Removed height and width concepts where possible
- Switched from shape checking decorator to function
- Added shape validation also for masks and keypoints
- Splitted utils into checks and serial

### CODE REVIEW

**Comment by soupault** [comment]
@lext this is ready for review.
This PR also fixes CutOut tests to solve the CI.

File: solt/core/_base_transforms.py [hunk 1]
```diff
@@ -541,8 +539,13 @@ def _apply_img(self, img: np.ndarray, settings: dict):
             Output Image
 
         """
+        ensure_valid_image(img, num_dims=(2,))
```
Comment (lext) [inline solt/core/_base_transforms.py:542]: This is done in a wrong place. The decorator has been checking that the output of the method returns a 3-dim shape.

File: solt/core/_base_transforms.py [hunk 2]
```diff
@@ -613,4 +617,4 @@ def _apply_pts(self, pts: Keypoints, settings: dict):
         pts_data[:, 0] /= pts_data[:, 2]
         pts_data[:, 1] /= pts_data[:, 2]
 
-        return Keypoints(pts_data[:, :-1], frame=(h_new, w_new))
+        return Keypoints(pts_data[:, :-1], frame=copy.copy(self.state_dict["frame_new"]))
```
Comment (lext) [inline solt/core/_base_transforms.py:620]: Copy is not needed I think

File: solt/core/_data.py [hunk 3]
```diff
@@ -283,8 +283,8 @@ def to_torch(self, as_dict=False, scale_keypoints=True, normalize=False, mean=No
             elif f == "P":
                 landmarks = torch.from_numpy(el.data).float()
                 if scale_keypoints:
-                    landmarks[:, 0] /= el.width - 1
-                    landmarks[:, 1] /= el.height - 1
+                    landmarks[:, 0] /= el.frame[1] - 1
```
Comment (lext) [inline solt/core/_data.py:286]: Should this already be refactored for nd-compatibility? It seems to be easy to modify...

File: solt/core/_data.py [hunk 4]
```diff
@@ -362,21 +362,15 @@ class Keypoints(object):
     Parameters
     ----------
     pts : numpy.ndarray
-        Key points as an numpy.ndarray in (x, y) format.
-    frame : (n, ) list-like
+        Array of keypoints. If 2D, has to be in (x, y) format.
+    frame : (n, ) list-like of ints
```
Comment (lext) [inline solt/core/_data.py:366]: SOLT uses tuples by default for everything

File: solt/transforms/_transforms.py [hunk 5]
```diff
@@ -17,13 +17,14 @@
...
     ALLOWED_CROPS,
     ALLOWED_INTERPOLATIONS,
-    ALLOWED_PADDINGS,
     ALLOWED_GRIDMASK_MODES,
 )
 from ..core import Stream
 from ..core import DataContainer, Keypoints
 from ..utils import (
-    img_shape_checker,
+    ensure_valid_image,
```
Comment (lext) [inline solt/transforms/_transforms.py:25]: What is the essential difference between an image and the mask? I think that image has always 3 dimensions in the current implementation, and masks have always 2 dimensions

File: solt/transforms/_transforms.py [hunk 6]
```diff
@@ -63,24 +64,28 @@ def _apply_img(self, img: np.ndarray, settings: dict):
             return np.ascontiguousarray(img[::-1, ::-1, ...])
 
     def _apply_mask(self, mask: np.ndarray, settings: dict):
+        ensure_valid_mask(mask, num_dims=(2, ))
```
Comment (lext) [inline solt/transforms/_transforms.py:67]: As mentioned previous, here and elsewhere, the check must be done only in the end, to make sure that the transform returns a correct shape (3 dims for images)

File: solt/transforms/_transforms.py [hunk 7]
```diff
@@ -1122,17 +1137,19 @@ def sample_transform(self, data):
...
-    @img_shape_checker
     def _apply_img(self, img: np.ndarray, settings: dict):
+        ensure_valid_image(img, num_dims=(2,))
+        shape_in = img.shape
+
         if self.blur == "g":
             return cv2.GaussianBlur(
                 img, ksize=(self.state_dict["k_size"], self.state_dict["k_size"]), sigmaX=self.state_dict["sigma"],
-            )
+            ).reshape(shape_in)
```
Comment (lext) [inline solt/transforms/_transforms.py:1147]: This reshape is not needed here given you will do the needed reshaping before returning the result.

File: solt/transforms/_transforms.py (hunk 5 above)
Comment (soupault) [reply solt/transforms/_transforms.py:25]: Yes, the main difference is that `image` is (d0, d1, ..., dn, ch)`, while `mask` is (d0, d1, ..., dn)`. Still, since these are different types, I believe the check has to be also in different functions. E.g. for mask, it may be reasonable to check whether it is binary, etc.

File: solt/transforms/_transforms.py (hunk 6 above)
Comment (soupault) [reply solt/transforms/_transforms.py:67]: This is a completely different check and is related to compatibility of the transforms with the different input types.
I removed the one implemented previously by the decorator, and implemented it in-place for the relevant functions (see `.reshape(shape_in)`). The decorator was only fixing the implicit convertion from `(h, w, 1)` to `(h, w)` done by multiple OpenCV functions.
I think the output reshaping has to be done explicitly within every transform.

File: solt/transforms/_transforms.py (hunk 7 above)
Comment (soupault) [reply solt/transforms/_transforms.py:1147]: See the comment above. It is needed, since OpenCV is corrupting the shape for the single-channel images.

File: solt/core/_data.py (hunk 4 above)
Comment (soupault) [reply solt/core/_data.py:366]: Tuple is list-like. This is solely for user convenience, so that the user does not need to worry whether to pass list or tuple. Internally, the type of `frame` is consistent.

File: solt/transforms/_transforms.py (hunk 5 above)
Comment (lext) [reply solt/transforms/_transforms.py:25]: I think it is best if the checks will be related to shapes only. the binary etc checks is the logic of the transform itself. I suggest to make one fucntion `ensure_shape(x, ndim)` that will simply reshape the tensor as needed. I don't think you need to check the incoming shape of the transform at all 🤔

File: solt/transforms/_transforms.py (hunk 6 above)
Comment (lext) [reply solt/transforms/_transforms.py:67]: I do not agree. From the development API point of view, it is much easier to have a function that will make sure that the shape is exactly as it should be. That function will be applied only to the output.

File: solt/transforms/_transforms.py (hunk 7 above)
Comment (lext) [reply solt/transforms/_transforms.py:1147]: I think adds more complexity to the codestyle and it diminishes the ease of extendibility of SOLT.

File: solt/core/_data.py [hunk 8]
```diff
@@ -5,6 +5,10 @@
 from solt.utils import validate_parameter
 
 
+IMAGENET_MEAN = (0.485, 0.456, 0.406)
```
Comment (lext) [inline solt/core/_data.py:8]: Should we rather move constants to constants then?

File: solt/core/_data.py [hunk 9]
```diff
@@ -240,11 +244,11 @@ def to_torch(self, as_dict=False, scale_keypoints=True, normalize=False, mean=No
             Whether to return the result as a dictionary. If a single item is present, then the singular naming
             will be used. If plural, then the plural will be used. The items will be stored and
             sorted a similar manner to the method ``from_dict``: images, masks, keypoints_array, and labels.
-            The same applies to a singular case/
+            The same applies to a singular case.
         scale_keypoints : bool
             Whether to scale keypoints to 0-1 range. ``True`` by default.
         normalize : bool
-            Whether to subtract mean
+            Whether to subtract `mean` and divide by `std`.
```
Comment (lext) [inline solt/core/_data.py:251]: Whether to subtract **the mean**

File: solt/transforms/_transforms.py [hunk 10]
```diff
@@ -643,46 +637,41 @@ def __init__(self, resize_to=None, interpolation="bilinear"):
...
-        else:
-            return cv2.resize(img, self.resize_to, interpolation=interp)
+        return cv2.resize(img, self.resize_to[::-1], interpolation=interp)
 
+    @ensure_valid_image(num_dims_spatial=(2,))
     def _apply_img(self, img: np.ndarray, settings: dict):
-        ensure_valid_image(img, num_dims=(2,))
         return self._apply_img_or_mask(img, settings)
 
+    @ensure_valid_image(num_dims_total=(2,))
```
Comment (lext) [inline solt/transforms/_transforms.py:655]: I suggest to remove this decorator for now as it does not generalize to 3D

File: solt/transforms/_transforms.py [hunk 11]
```diff
@@ -643,46 +637,41 @@ def __init__(self, resize_to=None, interpolation="bilinear"):
...
 
-        resize_x, resize_y = self.resize_to
-        # TODO: double check the notation
-        scale_x = resize_x / pts.frame[1]
-        scale_y = resize_y / pts.frame[0]
+        resize_d0, resize_d1 = self.resize_to
 
-        pts_data[:, 0] *= scale_x
-        pts_data[:, 1] *= scale_y
+        scale_d0 = resize_d0 / pts.frame[0]
```
Comment (lext) [inline solt/transforms/_transforms.py:669]: Replace to a more generic case when resize_to has more than 2 dimensions

File: tests/test_transforms.py [hunk 12]
```diff
@@ -450,17 +450,17 @@ def test_resize_img_to_arbitrary_size(img, mask, resize_to):
     if isinstance(resize_to, int):
         resize_to = (resize_to, resize_to)
 
-    scale_x = resize_to[0] / img.shape[1]
-    scale_y = resize_to[1] / img.shape[0]
+    scale_d0 = resize_to[0] / img.shape[0]
```
Comment (lext) [inline tests/test_transforms.py:453]: Should this test be written already for ND case?
